sklearn/datasets/base.py

```diff
@@ -919,3 +920,34 @@ def _fetch_remote(remote, dirname=None):
                       "file may be corrupted.".format(file_path, checksum,
                                                       remote.checksum))
     return file_path
+
+
+def _refresh_cache(files, compress):
+    # REMOVE in v0.23
+    import joblib
+    msg = "sklearn.externals.joblib is deprecated in 0.21"
+    with warnings.catch_warnings(record=True) as warns:
+        data = tuple([joblib.load(f) for f in files])
+
+    refresh_needed = any([str(x.message).startswith(msg) for x in warns])
+
+    if refresh_needed:
+        try:
+            for value, path in zip(data, files):
+                joblib.dump(value, path, compress=compress)
+        except IOError:
+            message = ("This dataset will stop being loadable in scikit-learn "
+                       "version 0.23 because it references a deprecated "
+                       "import path. Consider removing the following files "
+                       "and allowing it to be cached anew:\n%s"
+                       % ("\n".join(files)))
+            warnings.warn(message=message, category=DeprecationWarning)
+
+        other_warns = [w for w in warns if not str(w.message).startswith(msg)]
+        for w in other_warns:
+            warnings.warn(message=w.message, category=w.category)
```

Comment:
Ideally you should also be using the original module and line number. Is there an alternative function for best issuing the warning?

Reply:
there's the `warn_explicit`, but it didn't actually end up showing the warning and I couldn't fix it in 5 minutes, so I switched back to `warn`.